geode-core/src/test/java/org/apache/geode/internal/cache/PartitionedRegionTest.java

```diff
@@ -439,6 +443,66 @@ public void generatePRIdShouldNotThrowNumberFormatExceptionIfAnErrorOccursWhileR
         .doesNotThrowAnyException();
   }
 
+  @Test
+  public void getDataRegionForWriteThrowsTransactionExceptionIfNotDataStore() {
+    PartitionedRegion spyPartitionedRegion = spy(partitionedRegion);
+
+    KeyInfo keyInfo = mock(KeyInfo.class);
+    when(keyInfo.getBucketId()).thenReturn(1);
+    doReturn(null).when(spyPartitionedRegion).getDataStore();
+
+    Throwable caughtException =
+        catchThrowable(() -> spyPartitionedRegion.getDataRegionForWrite(keyInfo));
+
+    assertThat(caughtException).isInstanceOf(TransactionException.class).hasMessage(
+        "PartitionedRegion Transactions cannot execute on nodes with local max memory zero");
+  }
+
+  @Test
+  public void getDataRegionForWriteThrowsTransactionDataRebalancedExceptionIfGetInitializedBucketThrowsForceReattemptException()
+      throws Exception {
+    PartitionedRegion spyPartitionedRegion = spy(partitionedRegion);
+
+    KeyInfo keyInfo = mock(KeyInfo.class);
+    Object key = new Object();
+    PartitionedRegionDataStore dataStore = mock(PartitionedRegionDataStore.class);
+    when(keyInfo.getBucketId()).thenReturn(1);
+    when(keyInfo.getKey()).thenReturn(key);
+    when(keyInfo.isCheckPrimary()).thenReturn(true);
+    doReturn(dataStore).when(spyPartitionedRegion).getDataStore();
+    doThrow(new ForceReattemptException("")).when(dataStore)
+        .getInitializedBucketWithKnownPrimaryForId(key, 1);
+    doReturn(mock(InternalDistributedMember.class)).when(spyPartitionedRegion).createBucket(1, 0,
+        null);
+
+    Throwable caughtException =
+        catchThrowable(() -> spyPartitionedRegion.getDataRegionForWrite(keyInfo));
+
+    assertThat(caughtException).isInstanceOf(TransactionDataRebalancedException.class)
+        .hasMessage("Transactional data moved, due to rebalancing.");
```

Comment:
Is it worth extracting this string to a constant in the `PartitionedRegion` class so that it can be consistently referred to?

Reply:
Good idea, will make the change next time.